x/erc20/keeper/ibc_callbacks_integration_test.go

```diff
@@ -51,49 +67,142 @@ var _ = Describe("Recovery: Performing an IBC Transfer", Ordered, func() {
+			pair, err := s.EvmosChain.App.(*app.Evmos).Erc20Keeper.RegisterCoin(s.EvmosChain.GetContext(), evmosMeta)
+			s.Require().NoError(err)
+
+			aevmosInitialBalance := s.EvmosChain.App.(*app.Evmos).BankKeeper.GetBalance(s.EvmosChain.GetContext(), receiverAcc, claimstypes.DefaultClaimsDenom)
+
+			// 1. Send aevmos from Evmos to Osmosis
+			s.SendAndReceiveMessage(s.pathOsmosisEvmos, s.EvmosChain, claimstypes.DefaultClaimsDenom, amount, receiver, sender, 1)
+
+			aevmosAfterBalance := s.EvmosChain.App.(*app.Evmos).BankKeeper.GetBalance(s.EvmosChain.GetContext(), receiverAcc, claimstypes.DefaultClaimsDenom)
+			s.Require().Equal(aevmosInitialBalance.Amount.Int64()-amount, aevmosAfterBalance.Amount.Int64())
+
+			// check ibc aevmos coins balance on Osmosis
+			aevmosIBCBalanceBefore := s.IBCOsmosisChain.GetSimApp().BankKeeper.GetBalance(s.IBCOsmosisChain.GetContext(), senderAcc, aevmosIbcdenom)
+			s.Require().Equal(amount, aevmosIBCBalanceBefore.Amount.Int64())
+
+			// 2. Send aevmos IBC coins from Osmosis to Evmos
+			ibcCoinMeta := fmt.Sprintf("%s/%s", aevmosDenomtrace.Path, aevmosDenomtrace.BaseDenom)
+			s.SendBackCoins(s.pathOsmosisEvmos, s.IBCOsmosisChain, aevmosIbcdenom, amount, sender, receiver, 1, ibcCoinMeta)
+
+			// check ibc aevmos coins balance on Osmosis - should be zero
+			aevmosIBCSenderFinalBalance := s.IBCOsmosisChain.GetSimApp().BankKeeper.GetBalance(s.IBCOsmosisChain.GetContext(), senderAcc, aevmosIbcdenom)
+			s.Require().Equal(int64(0), aevmosIBCSenderFinalBalance.Amount.Int64())
+
+			// check aevmos balance after transfer - should be equal to initial balance
+			aevmosFinalBalance := s.EvmosChain.App.(*app.Evmos).BankKeeper.GetBalance(s.EvmosChain.GetContext(), receiverAcc, claimstypes.DefaultClaimsDenom)
+			s.Require().Equal(aevmosInitialBalance.Amount.Int64(), aevmosFinalBalance.Amount.Int64())
+
+			// check IBC Coin balance - should be zero
+			ibcCoinsBalance := s.EvmosChain.App.(*app.Evmos).BankKeeper.GetBalance(s.EvmosChain.GetContext(), receiverAcc, aevmosIbcdenom)
+			s.Require().Equal(int64(0), ibcCoinsBalance.Amount.Int64())
+
+			// Check ERC20 balances - should be zero
+			balanceTokenAfter :=
+				s.EvmosChain.App.(*app.Evmos).Erc20Keeper.BalanceOf(s.EvmosChain.GetContext(), contracts.ERC20MinterBurnerDecimalsContract.ABI, pair.GetERC20Contract(), common.BytesToAddress(receiverAcc.Bytes()))
+			s.Require().Equal(int64(0), balanceTokenAfter.Int64())
 		})
 		It("should transfer and convert original erc20", func() {
+			uosmoInitialBalance := s.IBCOsmosisChain.GetSimApp().BankKeeper.GetBalance(s.IBCOsmosisChain.GetContext(), senderAcc, "uosmo")
+
+			// 1. Send 'uosmo' from Osmosis to Evmos
+			s.SendAndReceiveMessage(s.pathOsmosisEvmos, s.IBCOsmosisChain, "uosmo", amount, sender, receiver, 1)
+
+			// 2. Transfer back the erc20 from Evmos to Osmosis
+			ibcCoinMeta := fmt.Sprintf("%s/%s", uosmoDenomtrace.Path, uosmoDenomtrace.BaseDenom)
```

Comment:
what is this metadata for?

Reply:
We need this to create properly the corresponding [`FungibleTokenPacketData`](https://github.com/cosmos/ibc-go/blob/b1f494c643280c73f3696461abbb11a39303b190/modules/apps/transfer/types/packet.pb.go#L28-L39) for IBC coins ([ADR-001](https://github.com/cosmos/ibc-go/blob/main/docs/architecture/adr-001-coin-source-tracing.md)). It is required to pass IBC coins denom in the format of `transfer/channelToA/denom` instead of the hash. Quoting the ADR-001:
>One final remark is that the FungibleTokenPacketData will remain the same, i.e with the prefixed full denomination, since the receiving chain may not be an SDK-based chain.